### DISCUSSION TITLE
Codespace with Java Swing and Desktop Lite, not working

### DISCUSSION THREAD

**Discussion by Mr-Coxall** [community/community#118415] [Programming Help] [answered]
### Body

I am trying to create a template repo so that I can do Java Swing development with my students. I am using the base image ([mcr.microsoft.com/devcontainers/universal:2](mcr.microsoft.com/devcontainers/universal:2)) and then I have added the Desktop Lite feature ([https://github.com/devcontainers/features/tree/main/src/desktop-lite](https://github.com/devcontainers/features/tree/main/src/desktop-lite)).
This way port 6080 (by default) is opened up and you can VNS into the "Desktop" and see the Swing GUI.

All the Desktop Lite processes seemingly work, but my Java Swing GUI elements do not show up on the screen inside the VNC desktop. I just get a "yellow" background. I know my Java Swing code is correct, here it is running in Repl.it ([https://replit.com/@JaneSmith2/Java-Swing-Test#src/main/java/Main.java](https://replit.com/@JaneSmith2/Java-Swing-Test#src/main/java/Main.java)).

Here is my GitHub repo: [https://github.com/Jane-Smith-2/Java-Swing-Test](https://github.com/Jane-Smith-2/Java-Swing-Test)

Here is my devcontainer.json:
`
{
  "image": "mcr.microsoft.com/devcontainers/universal:2",
    "features": {
		"ghcr.io/devcontainers/features/desktop-lite:1": {
			"password": "vscode",
			"webPort": "6080",
			"vncPort": "5901"
		}
	},
	"customizations": {
		"vscode": {
			"extensions": [
				"vscjava.vscode-java-pack"
			]
		}
	}
}
`

Thanks, Patrick

<img width="958" alt="Java_Swing" src="https://github.com/community/community/assets/3744969/66d70cfd-4272-47cc-862e-9609105faff6">

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by jfullstackdev** [comment, marked as answer]
answer from Copilot, maybe of help

--- 

The issue you're experiencing might be due to the lack of a proper window manager in the container. The Desktop Lite feature provides a basic X11 environment, but it doesn't include a window manager by default, which is often needed for Java Swing applications to display correctly.

You can try installing a lightweight window manager like `openbox` in your container. Here's how you can modify your `devcontainer.json` to include a `Dockerfile` that installs `openbox`:

```json
{
  "build": {
    "dockerfile": "Dockerfile"
  },
  "features": {
    "ghcr.io/devcontainers/features/desktop-lite:1": {
      "password": "vscode",
      "webPort": "6080",
      "vncPort": "5901"
    }
  },
  "customizations": {
    "vscode": {
      "extensions": [
        "vscjava.vscode-java-pack"
      ]
    }
  }
}
```

Then, create a `Dockerfile` in the `.devcontainer` directory with the following content:

```Dockerfile
# Start from the universal image
FROM mcr.microsoft.com/devcontainers/universal:2

# Install openbox
RUN apt-get update && export DEBIAN_FRONTEND=noninteractive \
    && apt-get -y install --no-install-recommends openbox

# Set openbox to start automatically
RUN echo 'openbox-session &' >> ~/.xinitrc
```

This will install `openbox` and configure it to start automatically when the X11 server starts. Now, when you open the VNC session, it should start `openbox` and your Java Swing application should display correctly.

Please note that this is a basic setup and might need further adjustments depending on your specific needs.

**Reply by Mr-Coxall** [reply]
Works great.
Thanks

**Comment by lynguyen11** [comment]
Thanks for the advice.  Because of this, I got it to work and I made a tutorial to get Java Swing with Graphics working for my students and other teachers. I hope this helps!
https://apps.mvhs.io/resources/codespaces/
